src/shared/firewall-util-nft.c

```diff
@@ -1105,3 +1151,143 @@ int fw_nftables_add_local_dnat(
+
+        NFTSetContext *c;
+        c = reallocarray(*s, *n + 1, sizeof(NFTSetContext));
+        if (!c)
+                return -ENOMEM;
+
+        *s = c;
+
+        c[(*n) ++] = (NFTSetContext) {
+                .nfproto = nfproto,
+                .table = TAKE_PTR(table_dup),
+                .set = TAKE_PTR(set_dup),
+        };
+
+        return 0;
+}
+
+int config_parse_nft_set_context(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                NFTSetContext **nft_set_context,
+                size_t *n) {
+        _cleanup_free_ char *family_str = NULL, *table = NULL, *set = NULL;
+        int nfproto, r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(nft_set_context);
+
+        if (isempty(rvalue)) {
+                *nft_set_context = nft_set_context_free_many(*nft_set_context, n);
+
+                return 0;
+        }
+
+        for (const char *p = rvalue;;) {
+                r = extract_many_words(&p, ":" WHITESPACE, EXTRACT_CUNESCAPE, &family_str, &table, &set, NULL);
+                if (r == -ENOMEM)
+                        return log_oom();
+                if (r == 0)
+                        return 0;
+                if (r != 3) {
+                        _cleanup_free_ char *esc = NULL;
+
+                        esc = cescape(rvalue);
+                        return log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse IPvxNFT set %s, ignoring", strna(esc));
+                }
+
+                assert(family_str);
+                assert(table);
+                assert(set);
+
+                nfproto = nfproto_from_string(family_str);
+                if (nfproto < 0) {
+                        _cleanup_free_ char *esc = NULL;
+
+                        esc = cescape(family_str);
+                        return log_syntax(unit, LOG_WARNING, filename, line, 0, "Unknown NFT protocol family %s, ignoring", strna(esc));
+
+                }
+                family_str = mfree(family_str);
+
+                if (!nft_identifier_good(table)) {
+                        _cleanup_free_ char *esc = NULL;
+
+                        esc = cescape(table);
+                        return log_syntax(unit, LOG_WARNING, filename, line, 0, "Invalid table name %s, ignoring", strna(esc));
+                }
+
+                if (!nft_identifier_good(set)) {
+                        _cleanup_free_ char *esc = NULL;
+
+                        esc = cescape(set);
+                        return log_syntax(unit, LOG_WARNING, filename, line, 0, "Invalid set name %s, ignoring", strna(esc));
+                }
+
+                NFTSetContext *c;
+                c = reallocarray(*nft_set_context, *n + 1, sizeof(NFTSetContext));
```

Comment:
GREEDY_REALLOC

Reply:
There's no GREEDY_REALLOCARRAY. Multiplication (size * n) can be done to use GREEDY_REALLOC but there's a theoretical possibility of overflow.

Reply:
Actually, there's no need for GREEDY_REALLOCARRAY since greedy_realloc() can be used instead and it uses size_multiply_overflow().

Reply:
Now with greedy_realloc()